flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaSerializationSchema.java

```diff
@@ -166,6 +167,15 @@ public void setPartitions(int[] partitions) {
 
     @Override
     public String getTargetTopic(RowData element) {
+        // no explicit topic defined in table options, use the writable metadata column topic
+        topic =
```

Comment:
What happens / should happen if the topic returned by metadata differs between different elements? We're now essentially reading the metadata of whichever element comes first and then ignore it forever, right? That isn't the behavior I would expect given the documentation.

Reply:
If the topic returned by metadata differs between different elements, the corresponding record could produced to different topic. I don't understand where isn't the behavior I would expect given the documentation.

Reply:
When `getTargetTopic(firstElement)` is called and `topic == null`, this will assign the value from the metadata to the topic. Now if `getTargetTopic(secondElement)` is called, `topic` will no longer be `null` and thus the metadata will not be read from the second element, so no matter what the metadata of this second element is, `topic` will be that of the first element still. Is that correct?

Reply:
@Airblader , yes, your point is correct. I would update the commit for this.